### ISSUE TITLE
Unit tests fails for export/import model

### ISSUE THREAD

**Issue by ilyatregubov** [moodlehq/moodle-mlbackend-python#49]
Hi,

if you run Moodle prediction unit tests like `vendor/bin/phpunit -v --filter core_analytics_prediction_testcase` there will be several errors/failures (tested with docker images 2.6.4 and 3.0.2 from https://hub.docker.com/r/moodlehq/moodle-mlbackend-python/tags), the latest succesfull run is for 2.5.0

Here are the logs: 

`Moodle 3.11.5+ (Build: 20220303), 9c6d7dbf7bc7d27c7afafbdf38f64cbbd4701b8d
Php: 7.4.28, pgsql: 10.18 (Ubuntu 10.18-0ubuntu0.18.04.1), OS: Linux 5.4.0-100-generic x86_64
PHPUnit 9.5.13 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.28
Configuration: /home/ilyatregubov/moodles/integration_311/moodle/phpunit.xml

.....F.E.........F.....E..                                        26 / 26 (100%)

Time: 05:57.047, Memory: 293.00 MB

There were 2 errors:

1) core_analytics_prediction_testcase::test_ml_export_import with data set "case-\mlbackend_python\processor-server" ('\mlbackend_python\processor', array(array(true, 'localhost', 5000, false, 'default', 'sshhhh')))
moodle_exception: Server error HTTP status code 500: 1

/home/ilyatregubov/moodles/integration_311/moodle/lib/mlbackend/python/classes/processor.php:642
/home/ilyatregubov/moodles/integration_311/moodle/lib/mlbackend/python/classes/processor.php:401
/home/ilyatregubov/moodles/integration_311/moodle/analytics/classes/model_config.php:91
/home/ilyatregubov/moodles/integration_311/moodle/analytics/classes/model.php:1605
/home/ilyatregubov/moodles/integration_311/moodle/analytics/tests/prediction_test.php:392
/home/ilyatregubov/moodles/integration_311/moodle/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///home/ilyatregubov/moodles/integration_311/moodle/vendor/phpunit/phpunit/phpunit:97

2) core_analytics_prediction_testcase::test_ml_evaluation_trained_model with data set "case-\mlbackend_python\processor-server" ('\mlbackend_python\processor', array(array(true, 'localhost', 5000, false, 'default', 'sshhhh')))
moodle_exception: Server error HTTP status code 500: 1

/home/ilyatregubov/moodles/integration_311/moodle/lib/mlbackend/python/classes/processor.php:642
/home/ilyatregubov/moodles/integration_311/moodle/lib/mlbackend/python/classes/processor.php:401
/home/ilyatregubov/moodles/integration_311/moodle/analytics/classes/model_config.php:91
/home/ilyatregubov/moodles/integration_311/moodle/analytics/classes/model.php:1605
/home/ilyatregubov/moodles/integration_311/moodle/analytics/tests/prediction_test.php:684
/home/ilyatregubov/moodles/integration_311/moodle/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///home/ilyatregubov/moodles/integration_311/moodle/vendor/phpunit/phpunit/phpunit:97

--

There were 2 failures:

1) core_analytics_prediction_testcase::test_ml_training_and_prediction with data set "quarters-\mlbackend_python\processor-server" ('\core\analytics\time_splittin...arters', 3, 4, '\mlbackend_python\processor', array(array(true, 'localhost', 5000, false, 'default', 'sshhhh')))
Failed asserting that '0' matches expected 1.

/home/ilyatregubov/moodles/integration_311/moodle/analytics/tests/prediction_test.php:254
/home/ilyatregubov/moodles/integration_311/moodle/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///home/ilyatregubov/moodles/integration_311/moodle/vendor/phpunit/phpunit/phpunit:97

2) core_analytics_prediction_testcase::test_ml_multi_classifier with data set "notimesplitting-\mlbackend_python\processor-server" ('\core\analytics\time_splittin...itting', '\mlbackend_python\processor', array(array(true, 'localhost', 5000, false, 'default', 'sshhhh')))
Failed asserting that '2' matches expected 0.

/home/ilyatregubov/moodles/integration_311/moodle/analytics/tests/prediction_test.php:578
/home/ilyatregubov/moodles/integration_311/moodle/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///home/ilyatregubov/moodles/integration_311/moodle/vendor/phpunit/phpunit/phpunit:97

ERRORS!
Tests: 26, Assertions: 215, Errors: 2, Failures: 2.
`

**Comment by ilyatregubov** [comment]
A bit more details. I ran unit tests locally, not using container and have some extra info:

`  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/export.py", line 22, in <module>
    export_classifier()
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/export.py", line 15, in export_classifier
    exportdir = classifier.export_classifier(sys.argv[3])
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/processor/estimator.py", line 588, in export_classifier
    sess = classifier.get_session()
AttributeError: 'TF' object has no attribute 'get_session'
`

So seems that that is due to tensorflow upgrade and get_session is deprecated or smth?

**Comment by douglasbagnall** [comment]
> So seems that that is due to tensorflow upgrade and get_session is deprecated or smth?

Yes, I think that's the problem.

**Comment by ilyatregubov** [comment]
I was trying to use a workaround from https://stackoverflow.com/questions/58255821/how-to-use-k-get-session-in-tensorflow-2-0-or-how-to-migrate-it/58442121#58442121, so have 

`sess = tf.compat.v1.keras.backend.get_session()`

But it error in other place then with

`Runtime:       PHP 7.4.28
Configuration: /home/ilyatregubov/moodles/integration_master/moodle/phpunit.xml

.Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/import.py", line 24, in <module>
    import_classifier()
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/import.py", line 17, in import_classifier
    classifier.import_classifier(sys.argv[3])
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/processor/estimator.py", line 619, in import_classifier
    initial_weights=import_vars)
  File "/home/ilyatregubov/.local/lib/python3.7/site-packages/moodlemlbackend/processor/estimator.py", line 243, in get_classifier
    n_features = X.shape[1]
AttributeError: 'bool' object has no attribute 'shape'
E                                                                  2 / 2 (100%)
`

**Comment by douglasbagnall** [comment]
> > So seems that that is due to tensorflow upgrade and get_session is deprecated or smth?
> 
> Yes, I think that's the problem.

Well, the bigger problem is that the local interface (as opposed to the web interface) is entirely untested on the Python side, so it is likely broken in many small ways.

Also, import and export are fairly untested on the web side, as I couldn't work out what they were actually useful for.

**Comment by douglasbagnall** [comment]
From the original report
```
core_analytics_prediction_testcase::test_ml_export_import with data set "case-\mlbackend_python\processor-server" ('\mlbackend_python\processor', array(array(true, 'localhost', 5000, false, 'default', 'sshhhh')))
moodle_exception: Server error HTTP status code 500: 1
```
it looks like it the test is running the web server, which means it is likely to have written to a log *somewhere*.

Exactly where it wrote is hard to determine, and it tends to skip around, as various operations reset it (why we do this I don't know). Look for a file called `info.log`, at the end of a path like `{40 hexdigits} / {40 other hexdigits} /logs/ {unix epoch timestamp (16...)}  /info.log`.

**Comment by ilyatregubov** [comment]
Hi Douglas,

I am pretty sure that that 500 error refers to get_session error I posted. So first I ran into that 500 error when using container. Then I tried it locally and I got that error. I think docker just couln't compose the error response properly and gave generic error code. You can reproduce the error if you install mlbackend on local machine. Then I tried to fix local version with `sess = tf.compat.v1.keras.backend.get_session()` and ran into another error I mentioned.

So to reproduce - install mlbackend locally with pip, set `$CFG->pathtopython = '/usr/bin/python3.7';` and then run `vendor/bin/phpunit -v --filter core_analytics_prediction_testcase`

**Comment by ilyatregubov** [comment]
There are several commits that are causing different problem. There has been 2 merges catalyst-catalyst-3.0 and catalyst-2.5. So thats what I found so far:

1. Lets start with first merge of catalyst-2.5 branch i.e. revert to `d7907ffd0287ff7b18847d6163668f27b51b2021`. To use that version you need to have use numpy 1.16.5 and tensorflow 1.15.5.Import unit tests are failing there already with this error `AttributeError: 'bool' object has no attribute 'shape'` But I was able to get them all passing by reverting these two commits `b0f5294e0bf3fc9f3ff731e3f25b747e61aa80eb training: respect previously found invariant columns` and `faa5172c63268a087044152e73e4cf63e56bb6a1 Classifier: remove invariant columns` After that all unit tests (not just import - all related to analytics) were run ok for catalyst-2.5 merge 
2. Now lets run those tests on latest master. So you will need to have tensorflow 2.4.2 and numpy 1.19.5. Checkout master and revert those two commits from 1. Now since we are using tf2 we have several deprecations to handle. So we need to substitute `sess = classifier.get_session()` with `sess = tf.compat.v1.keras.backend.get_session()` and `import tensorflow as tf` with `import tensorflow.compat.v1 as tf`. Once you done with those changes and install resulting mlbackend package run unite tests. Import unit tests will be passing. But we will still have several failures like `1) core_analytics_prediction_testcase::test_ml_training_and_prediction with data set "no_splitting-\mlbackend_python\processor" ('\core\analytics\time_splittin...itting', 0, 1, '\mlbackend_python\processor', null)
Failed asserting that '1' matches expected 0.` I have no idea how to fix it, but I am pretty sure this commit is responsible `384552db2b978b1076bece12b9651d760f3a7ba4 Switch to Keras based model` Its failing right after it.

So to sum up - several unit tests are failing. There are 2 commits on catalyst-2.5 that seem to cause import unit test to fail. The rest is due to `Switch to Keras based model` which looks like a big change

**Comment by danmarsden** [comment]
thanks @ilyatregubov - as mentioned in chat, I think we need to get Unit tests inside the mlbackend repo that match the import/export test in Moodle - We (Douglas) added a lot of new unit tests in the mlbackend code we sent through, but it would be good to add more so that we are covered well.

**Comment by stronk7** [comment]
Can this now (with #51 merged and images rebuilt) closed? Do we need some other issue about the missing tests or this one is ok (keeping it open).

**Comment by danmarsden** [comment]
yeah - this one should be able to be closed now, Douglas included a bunch of new unit tests in the mlbackend, including some with small/trival data-sets similar to the ones that Moodle was trying to train/predict on. They aren't exactly the same as the ones from Moodle, but I think the coverage looks ok to me.

**Comment by ilyatregubov** [comment]
Closing - thanks all
